tools/mod123encry/Decrypt.cpp

```diff
@@ -0,0 +1,122 @@
+        if (memcmp(&data[i], pattern1, sizeof(pattern1)) == 0
+            && memcmp(&data[i + sizeof(pattern1) + 4], pattern2, sizeof(pattern2)) == 0
+            && memcmp(&data[i + sizeof(pattern1) + sizeof(pattern2) + 8], pattern3, sizeof(pattern3)) == 0) {
+            return i;
+        }
+    }
+    return data.size();
+}
+
+u32 Decryptor::DoDecryptLvl2(u32 tableOffset) {
+    u32 *pool = (u32 *)&data[tableOffset + 104];
+    pool[1] -= info.start + info.size + 0x1300;
+    pool[2] -= info.start + info.size + 0x1300;
+    pool[3] -= 0x1300;
+    u32 param = pool[2];
+    u32 start = pool[3];
+    u32 size = pool[1];
+    u32 keys[4] = {
+        size ^ param,
+        size ^ ((param << 8) | (param >> 24)),
+        size ^ ((param << 16) | (param >> 16)),
+        size ^ ((param << 24) | (param >> 8)),
+    };
+    CryptoRC4Context buffer((const u8 *)keys);
+    buffer.Decrypt((u32 *) &data[start - info.start], (u32 *) &data[start + size - info.start]);
+    return tableOffset + 120;
+}
+
+void Decryptor::DecryptLvl2() {
+    u32 i = 0;
+    while ((i = FindDecryLvl2(i)) != data.size()) {
+        i = DoDecryptLvl2(i);
+    }
+}
+
+u32 Decryptor::DoDecryptLvl1(u32 tableOffset) {
+    FATEntry *table;
+    FATEntry *table_start = (FATEntry *)(&data[tableOffset]);
+    for (table = table_start; table->start != 0 && table->end != 0; table++) {
+        table->start -= 0x1300;
+        table->end -= info.start + info.size + 0x1300;
+        u32 start_offs = table->start - info.start;
+        u32 size = table->end & ~3;
+        for (int i = start_offs; i < start_offs + size; i += 4) {
+            u32 & word = (u32 &)data[i];
+            switch (GetInsnType(word)) {
+            case 1:
+            case 2:
+                word = (((word & ~0xFF000000) - 0x1300) & ~0xFF000000) | ((word & 0xFF000000) ^ 0x01000000);
+                break;
+            case 3:
+                word = (((word & ~0xFF000000) - 0x4C2) & ~0xFF000000) | ((word & 0xFF000000) ^ 0x01000000);
+                break;
+            default:
+                u8 *ptr = (u8 *)&word;
+                word = ((ptr[0] ^ 0x56) << 0) | ((ptr[1] ^ 0x65) << 8) | ((ptr[2] ^ 0x56) << 16) | ((ptr[3] ^ 0xF0) << 24);
+                break;
+            }
+        }
+    }
+    return (u8 *)table - (u8 *)table_start + tableOffset + sizeof(*table);
+}
+
+void Decryptor::DecryptLvl1() {
+    for (int i = info.sinit_start; i != info.sinit_end; i += 4) {
+        if (*(u32 *)&data[i - info.start] != 0) {
+            (void)DoDecryptLvl1(*(u32 *)&data[i - info.start] - info.start + 16);
+        }
+    }
+}
+
+void Decryptor::Decrypt() {
+    DecryptLvl1();
+    DecryptLvl2();
+}
+
+void Decryptor::Write(std::ofstream &outfile) {
+    outfile.write((char *)data.data(), data.size());
+}
+
+DecryptOptions::DecryptOptions(char ** argv) {
+    baserom = new NtrRom(argv[2], std::ios::binary);
```

Comment:
if the filestream is implicitly deleted, why wouldn't the baserom be?

Reply:
my (iffy) understanding is that the filestream is created without using `new`, so it has an **automatic** storage duration and the destructor would be called automatically once the filestream goes out-of-scope.

but because the baserom was created with `new`, it has a **dynamic** storage duration, meaning it will not be automatically freed once it's out-of-scope, so we need to free it ourselves.